Cisco IOS Support

fbradyirl opened the pull request:
**Description:**
This adds support for using Cisco IOS ARP table for device tracking.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation:** home-assistant/home-assistant.github.io#1419

**Example entry for `configuration.yaml` (if applicable):**
```yaml
device_tracker:
  - platform: cisco_ios
    host: 10.10.10.7
    username: ciscossh
    password: cisco12345
```

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L16
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L51

fbradyirl commented:
question for you guys. @balloob @fabaff @robbiet480 

The Cisco IOS ARP table shows all devices with an 'age' field which I have seen anywhere from 0 to 250 minutes. 0 being the most recently interacted with devices. 

Should I add logic to say, dispose of anything older than 30 or 60 minutes, for example? Or does that logic sit outside the device_tracker component?

kellerza commented:
Stale devices are already [removed periodically](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/device_tracker/__init__.py#L163) if the device is not "seen" for the `consider_home` interval. 

Keeping that in mind, I'm not sure how accurate this will be with arp cache timeouts of the 250minutes you mention. If the arp cache only expires after 250 minutes, and you keep on "seeing" the device for that time, HASS will only mark it as stale 250min+`consider_home` interval. I'm pretty sure the clients on the network will have shorter arp cache timeouts, which might reset the IOS timer if it receives an arp request, but that would then be client dependent. You can also not just ignore age above a certain time, since the client might still be there and happily forwarding, without needing to refresh it's arp entry. If you can reduce the arp cache timeout on IOS it might help, so that an active arp request actually verifies the client's presence.

Please `git rebase` on the latest dev (and force update) to clean up your PR.

I see you have two dependencies for this component. Any chance you can reduce this requirement? Some other components already use `pexpect` for SSH interaction, which might work, but maybe you have a good reason for `netmiko`. However, converting the mac address format can easily be done with 1-2 lines of standard Python and really don't require another library. Google "left-pad" disaster to see what the impact of too many dependencies can be :-)

balloob commented:
Your PR changes 107 files. Please clean up your branch.

Commit:
Removed module requirements by using pexpect

fbradyirl commented:
@kellerza Thanks for your thoughts.

> Stale devices are already removed periodically if the device is not "seen" for the consider_home interval.

Great. That's good to know.

> Keeping that in mind, I'm not sure how accurate this will be with arp cache timeouts of the 250minutes you mention. If the arp cache only expires after 250 minutes, and you keep on "seeing" the device for that time, HASS will only mark it as stale 250min+consider_home interval. I'm pretty sure the clients on the network will have shorter arp cache timeouts, which might reset the IOS timer if it receives an arp request, but that would then be client dependent. You can also not just ignore age above a certain time, since the client might still be there and happily forwarding, without needing to refresh it's arp entry. 

Yes, this is exactly as I am seeing it during manual monitoring.

> If you can reduce the arp cache timeout on IOS it might help, so that an active arp request actually verifies the client's presence.

I will experiment with this.

> Please git rebase on the latest dev (and force update) to clean up your PR.

That's done now.

> I see you have two dependencies for this component. Any chance you can reduce this requirement? Some other components already use pexpect for SSH interaction, which might work, but maybe you have a good reason for netmiko. 

I have refactored now to use `pexpect` instead of `netmiko`, although I fear that my implementation may not suit [subtle variants in SSH](https://www.networking-forums.com/programming-goodies-and-software-defined-networking/interesting-new-paramiko-variant-(ssh-module-for-python)/) on different Cisco IOS devices and firmware versions. It supports lots of devices with the same python API. e.g:

```
Cisco IOS
Cisco IOS XE
Cisco IOS XR
Cisco ASA
Cisco NX-OS 
+ other vendors
```

As my script stands right now, it just supports my version of Cisco IOS, which is version 15. Maybe that is OK for most people though. 

> However, converting the mac address format can easily be done with 1-2 lines of standard Python and really don't require another library.

Yes, good point. I have also removed the `netaddr` requirement.

Commit:
Added requirements and coveragerc entry for cisco_ios

fbradyirl commented:
I have just tested setting the ARP cache timeout to 4 minutes (which forces IOS to do an ARP broadcast every 2 minutes) and it seems to be working well now. Here are the commands I used:

```
conf t
interface Vlan1
arp timeout 240
```
I can add that to the docs for the component.

kellerza commented on homeassistant/components/device_tracker/cisco_ios.py:
```diff
@@ -0,0 +1,164 @@
+            # show ip arp
+            # Protocol  Address | Age (min) | Hardware Addr | Type | Interface
+            lines_result.remove(lines_result[0])
+            lines_result.remove(lines_result[0])
```
You can probably combine this into a single line: `lines_result = lines_result[2:]`

kellerza commented on homeassistant/components/device_tracker/cisco_ios.py:
```diff
@@ -0,0 +1,164 @@
+                        age = int(age)
+                        if age < 1:
+                            _LOGGER.debug('ADDING: age < 1: ' + line)
+                            _LOGGER.debug('\t\t: ' + mac)
```
Please combine this into a single line using `%s` substitution: `_LOGGER.debug('Adding: age<1 %s %s', mac, line)`

kellerza commented on homeassistant/components/device_tracker/cisco_ios.py:
```diff
@@ -0,0 +1,164 @@
+                        if age < 1:
+                            _LOGGER.debug('ADDING: age < 1: ' + line)
+                            _LOGGER.debug('\t\t: ' + mac)
+                            last_results.append(mac.upper())
```
Maybe you can move this `.upper()` part to you custom mac parsing function?

kellerza commented on homeassistant/components/device_tracker/cisco_ios.py:
```diff
@@ -0,0 +1,164 @@
+                    # ['Internet', '10.10.11.1', '-', '0027.d32d.0123', 'ARPA',
+                    # 'GigabitEthernet0']
+                    _protocol, _address, age, hw_addr, _arptype, _interface = \
+                        line.split()
```
You are splitting the line twice here, an alternative could be:

```python
parts = line.split()
if len(parts) != 6:
  continue
age = parts[2]
hw_addr = parts[3]
```

kellerza commented on homeassistant/components/device_tracker/cisco_ios.py:
```diff
@@ -0,0 +1,164 @@
+        """Open connection to the router and get arp entries."""
+        from pexpect import pxssh
+        import re
+        _LOGGER.info('Checking ARP')
```
Do we really need to log here?

kellerza commented on homeassistant/components/device_tracker/cisco_ios.py:
```diff
@@ -0,0 +1,164 @@
+    00:1D:EC:02:07:AB
+
+    :param cisco_hwaddr: HWAddr String from Cisco ARP table
+    :return: Regular standard MAC address
```
Please remove these :param comments

kellerza reviewed (changes requested):
Unfortunately you squashed your previous commits, but getting the feeling that `paramiko` (or `netmiko`) is a fairly clean Python only way to interface with ssh compared to `pexpect`.

Totally up to you, but you are welcome to revert back to `**miko`

Commit:
Updates based on review comments

fbradyirl commented:
> Totally up to you, but you are welcome to revert back to **miko

If future cisco ios component users start to see issues with `pyexpect`, then I may revert it back to `miko`, but I guess it is safe to leave it for now.

Thanks for your feedback. I have addressed your review comments in the latest push.

kellerza commented:
Thank you! :dolphin:

fbradyirl commented:
Great thanks for the merge!
